### PR TITLE
[ENH] Hugging Face interface for pre-trained forecasters

### PR DESCRIPTION
**Pull request by benHeid**
#### Reference Issues/PRs
Towards #5790 

#### What does this implement/fix? Explain your changes.
This forecaster loads a pertained transformer model from huggingface

#### Does your contribution introduce a new dependency? If yes, which one?
The `transformers` library would be added

#### What should a reviewer concentrate their feedback on?

Is the translation from the `transformers` distribution to the `supro` distribution correct.

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [ ] Optionally, I've added myself and possibly others to the [CODEOWNERS](https://github.com/sktime/sktime/blob/main/CODEOWNERS) file - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [ ] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [ ] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [ ] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [ ] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

### CODE REVIEW

**Comment by fkiraly** [comment]
nice! short note, for tests to run, you need to add dependencies to `pyproject.toml`.

**Comment by benHeid** [comment]
> nice! short note, for tests to run, you need to add dependencies to `pyproject.toml`.

I know. I created the draft PR mainly for tracking the progress :)

**Comment by fkiraly** [comment]
ok. If you don't mind then, I'll kill all CI until the release to give priority to release facing CI.

**Comment by yarnabrina** [comment]
May I please request you to add some temporary debug commit to to disable CI parts you don't want to test? Currently any change in pyproject.toml triggers 200+ jobs (after code quality and soft dependency checks).

We'll revert that of course at the end and run full CI just before the merge.

File: sktime/forecasting/hf_transformers_forecaster.py [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+        if self.deterministic:
+            torch.manual_seed(42)
+            np.random.seed(42)
+
+        if fh is None:
+            fh = self.fh
+        fh = fh.to_relative(self.cutoff)
+
+        if min(fh._values) < 0:
+            raise NotImplementedError("LTSF is not supporting insample predictions.")
```
Comment (yarnabrina) [inline sktime/forecasting/hf_transformers_forecaster.py:226]: Typo in model name

File: sktime/forecasting/hf_transformers_forecaster.py [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+            compute_metrics=self._compute_metrics,
+            callbacks=self.callbacks
+        )
+        trainer.train()
+
+    def _predict(self, fh, X=None):
+        if self.deterministic:
+            torch.manual_seed(42)
+            np.random.seed(42)
+
```
[the remark this answers is not in the document]
Comment (yarnabrina) [reply sktime/forecasting/hf_transformers_forecaster.py:220]: Also, will it make sense to let the user specify the seed?

File: sktime/forecasting/hf_transformers_forecaster.py [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+    Parameters
+    ----------
+    model_path : str
+        Path to the huggingface model to use for forecasting
+    fit_strategy : str, default="minimal"
+        Strategy to use for fitting the model. Can be "minimal" or "full"
+    validation_split : float, default=0.2
+        Fraction of the data to use for validation
+    config : dict, default={}
+        Configuration to use for the model.
```
Comment (yarnabrina) [inline sktime/forecasting/hf_transformers_forecaster.py:31]: Where do we get the valid options? I don't think this is available for all models.

**Review by yarnabrina — COMMENTED** [review]
I added some comments seeing this as ready for review. If it's WIP, please ignore in that case.

File: sktime/forecasting/hf_transformers_forecaster.py (hunk 3 above)
Comment (benHeid) [reply sktime/forecasting/hf_transformers_forecaster.py:33]: This depends on the model. I added a statement that you should take a look at the model card.

File: sktime/forecasting/hf_transformers_forecaster.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (benHeid) [reply sktime/forecasting/hf_transformers_forecaster.py:224]: It is just for testing, thus I would not let the user specify the seed.

File: sktime/forecasting/hf_transformers_forecaster.py [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+        # Update config with user provided config
+        _config = config.to_dict()
+        _config.update(self.config)
+        _config["num_dynamic_real_features"] = X.shape[-1] if X is not None else 0
+        _config["num_static_real_features"] = 0
+        _config["num_dynamic_real_features"] = 0
+        _config["num_static_categorical_features"] = 0
+        _config["num_time_features"] = 0 if X is None else X.shape[-1]
+
+        del _config["feature_size"]
```
[the remark this answers is not in the document]
Comment (benHeid) [reply sktime/forecasting/hf_transformers_forecaster.py]: I added a check

File: sktime/forecasting/hf_transformers_forecaster.py [hunk 5]
```diff
@@ -0,0 +1,372 @@
+"""Adapter for using huggingface transformers for forecasting."""
+
+import numpy as np
+import pandas as pd
+import torch
+from torch.utils.data import Dataset
+from transformers import AutoConfig, Trainer, TrainingArguments
+
```
[the remark this answers is not in the document]
Comment (benHeid) [reply sktime/forecasting/hf_transformers_forecaster.py:23]: You are right.

File: sktime/forecasting/hf_transformers_forecaster.py [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+    Forecaster that uses a huggingface model for forecasting.
+
+    This forecaster fetches the model from the huggingface model hub.
+    Note, this forecaster is in an experimental state. It is currently only
+    working for Informer, Autoformer, and TimeSeriesTransformer. 
+
+    Parameters
+    ----------
+    model_path : str
+        Path to the huggingface model to use for forecasting
```
[the remark this answers is not in the document]
Comment (benHeid) [reply sktime/forecasting/hf_transformers_forecaster.py]: Currently, Informer, Autoformer and TimeSeriesModelTransformer are supported. I added a comment.

File: sktime/forecasting/hf_transformers_forecaster.py [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+        training_args = TrainingArguments(**training_args)
+
+        if self.fit_strategy == "minimal":
+            if len(info["mismatched_keys"]) == 0:
+                return  # No need to fit
+        elif self.fit_strategy == "full":
+            for param in self.model.parameters():
+                param.requires_grad = True
+        else:
+            raise Exception("Unknown fit strategy")
```
[the remark this answers is not in the document]
Comment (benHeid) [reply sktime/forecasting/hf_transformers_forecaster.py]: I agree.

File: sktime/forecasting/hf_transformers_forecaster.py [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+
+        trainer = Trainer(
+            model=self.model,
+            args=training_args,
+            train_dataset=train_dataset,
+            eval_dataset=eval_dataset,
+            compute_metrics=self._compute_metrics,
+            callbacks=self.callbacks
```
[the remark this answers is not in the document]
Comment (benHeid) [reply sktime/forecasting/hf_transformers_forecaster.py]: I think `metric_for_best_model` and `greater_is_better` are part of the training_args. Thus, you would need to pass the in the training_args dict.

File: sktime/forecasting/hf_transformers_forecaster.py (hunk 1 above)
Comment (benHeid) [reply sktime/forecasting/hf_transformers_forecaster.py]: Thanks, I changed it.

**Comment by benHeid** [comment]
FYI. I am currently not able to fix these issues, since I a cannot reproduce it on torch-nightly and on the most recent stable version a different bug related to Mac's M chip is raised... 
Thus, I will not be able to fix it until next week when having access to an other computer..

**Comment by fkiraly** [comment]
again those M chips...

**Comment by benHeid** [comment]
I was able to solve this by running it on the CPU instead of using the MPS. Thus, I would appreciate reviews on this PR. 

(Side Note: We are testing only on CPU, but I learned that PyTorch on CPU might behaves differently than on Cuda or MPS https://github.com/pytorch/pytorch/issues/123119)

File: pyproject.toml [hunk 9]
```diff
@@ -266,6 +266,7 @@ dl = [
   'FrEIA; python_version < "3.12"',
   'tensorflow<=2.14,>=2; python_version < "3.12"',
   'torch; python_version < "3.12"',
+  'transformers[torch]; python_version < "3.12"',
```
Comment (yarnabrina) [inline pyproject.toml:269]: According to this, it seems even 3.11 is not supported.

https://github.com/huggingface/transformers/blob/main/setup.py#L460

This may not have been updated though, I am not sure.

File: sktime/forecasting/hf_transformers_forecaster.py [hunk 10]
```diff
@@ -0,0 +1,398 @@
...
+    def _fit(self, y, X, fh):
+        # Load model and extract config
+        config = AutoConfig.from_pretrained(self.model_path)
+
+        # Update config with user provided config
+        _config = config.to_dict()
+        _config.update(self._config)
+        _config["num_dynamic_real_features"] = X.shape[-1] if X is not None else 0
+        _config["num_static_real_features"] = 0
+        _config["num_dynamic_real_features"] = 0
```
Comment (yarnabrina) [inline sktime/forecasting/hf_transformers_forecaster.py:137]: Not a review, just a question: what is th difference between `num_dynamic_real_features` and `num_time_features`?

File: sktime/forecasting/hf_transformers_forecaster.py [hunk 11]
```diff
@@ -0,0 +1,398 @@
...
+            train_dataset=train_dataset,
+            eval_dataset=eval_dataset,
+            compute_metrics=self._compute_metrics,
+            callbacks=self._callbacks,
+        )
+        trainer.train()
+
+    def _predict(self, fh, X=None):
+        if self.deterministic:
+            transformers.set_seed(42)
```
Comment (yarnabrina) [inline sktime/forecasting/hf_transformers_forecaster.py:245]: Is `transformers` imported at this point?

Comment (benHeid) [reply sktime/forecasting/hf_transformers_forecaster.py:245]: Yes since it is specified as a soft dependency of the Forecaster and if this soft-dependency exist then it is imported in line 21. And if the soft-dependency is not installed than it would raise an error before notifying the user that the soft-dependency is not installed.

File: sktime/forecasting/hf_transformers_forecaster.py (hunk 10 above)
Comment (benHeid) [reply sktime/forecasting/hf_transformers_forecaster.py:137]: To be honest. I am not quite sure:

This is the documentation from hugging face: https://huggingface.co/docs/transformers/model_doc/autoformer#transformers.AutoformerConfig

```

num_time_features (int, optional, defaults to 0) — The number of time features in the input time series.

num_dynamic_real_features (int, optional, defaults to 0) — The number of dynamic real valued features.

num_static_categorical_features (int, optional, defaults to 0) — The number of static categorical features.
```

Based on that I decided to just ignore num_dynamic_real_features and put everything into num_time_features

File: pyproject.toml (hunk 9 above)
Comment (benHeid) [reply pyproject.toml:269]: That is weird... At least locally I was able to install it with python 3.11 and in the CI it is installed too:

https://github.com/sktime/sktime/actions/runs/8516512860/job/23325741934?pr=5796

File: sktime/forecasting/hf_transformers_forecaster.py (hunk 11 above)
Comment (yarnabrina) [reply sktime/forecasting/hf_transformers_forecaster.py:245]: Sorry, I did miss those lines outside the estimator definition. I'm a bit concerned though about importing globally always if it's present. This means transformers and all its dependencies will be imported always even if I don't use anything other than ARIMA, right? It may have some unpleasant side effects, including higher load time etc.

Comment (benHeid) [reply sktime/forecasting/hf_transformers_forecaster.py:245]: Not sure if I understand this correctly. 

If you are not importing this forecaster it should not import transformers or?

File: pyproject.toml (hunk 9 above)
Comment (yarnabrina) [reply pyproject.toml:269]: Yes, I verified this over weekend as well for a personal project. Don't know why they haven't updated, but probably we can ignore.

Comment (yarnabrina) [inline pyproject.toml:269]: @fkiraly given your comment on the dask PR, I am now very confused how this is not failing. This PR shows all tests (so installations) as passing. 

(For context to @benHeid, it was suggested in #6282 that extra marker is required. )

Comment (fkiraly) [reply pyproject.toml:269]: I am as confused as you are. I thought this would not work, so why does it?

I explicitly checked that the `dl` dep set is actually used, and it is, in `test-full`.

**Comment by fkiraly** [comment]
@benHeid, is this ready to merge?

**Review by fkiraly — APPROVED** [review]
Mergeable, as discussed
